src/core/BufferAttribute.js

```diff
@@ -20,23 +20,79 @@ function BufferAttribute( array, itemSize, normalized ) {
 
 	this.name = '';
 
-	this.array = array;
-	this.itemSize = itemSize;
+	this._array = array;
+	this._itemSize = itemSize;
 	this.count = array !== undefined ? array.length / itemSize : 0;
-	this.normalized = normalized === true;
+	this._normalized = normalized === true;
 
 	this.usage = StaticDrawUsage;
 	this.updateRange = { offset: 0, count: - 1 };
 
 	this.version = 0;
+	this.versionVAO = 0;
 
 }
 
-Object.defineProperty( BufferAttribute.prototype, 'needsUpdate', {
+Object.defineProperties( BufferAttribute.prototype, {
 
-	set: function ( value ) {
+	needsUpdate: {
 
-		if ( value === true ) this.version ++;
+		set: function ( value ) {
+
+			if ( value === true ) this.version ++;
+
+		}
+
+	},
+
+	array: {
```

Comment:
@mrdoob If we consider properties like `array`, `itemSize`, `normalized` as read-only properties, this PR could be simplified at certain places. Is there a use case for changing `itemSize` or `normalized` of an existing buffer attribute after initial creation? Besides, the engine does not provide `setArray()` anymore so it assume buffer attributes to be fixed size. Meaning `array` could be declared as read-only, too.

Reply:
Similarly can we consider `meshPerAttribute` of `InstancedBufferAttribute` and `InstancedInterleavedBufferAttribute` as read-only? If we can, we can further simplify the code.

I don't think users want to change it on the fly (in the practical use). And `WebGLRenderer` seems to already consider it as immutable, changing `meshPerAttribute` doesn't have an effect to `geometry._maxInstanceCount` in the following code

https://github.com/mrdoob/three.js/blob/dev/src/renderers/WebGLRenderer.js#L938-L942

```javascript
if ( geometry._maxInstanceCount === undefined ) {

	geometry._maxInstanceCount = data.meshPerAttribute * data.count;

}
```

Reply:
I think all properties of buffer attributes which values are set via ctor parameters should be read-only. I'm not aware of a compelling use case for changing them.